test/replication-luatest/linearizable_test.lua

```diff
@@ -0,0 +1,314 @@
+local t = require('luatest')
+local misc = require('test.luatest_helpers.misc')
+local cluster = require('test.luatest_helpers.cluster')
+local server = require('test.luatest_helpers.server')
+local proxy = require('test.luatest_helpers.proxy.proxy')
+local g = t.group('linearizable-read')
+
+local fiber = require('fiber')
+
+local function build_replication(num_instances)
+    local t = {}
+    for i = 1, num_instances do
+        table.insert(t, server.build_instance_uri('server_' .. i .. '_proxy'))
+    end
+    return t
+end
+
+local num_servers = 3
+
+g.before_all(function()
+    g.cluster = cluster:new({})
+    g.servers = {}
+    g.box_cfg = {
+        replication = build_replication(num_servers),
+        replication_timeout = 0.1,
+        memtx_use_mvcc_engine = true,
+    }
+    g.servers[1] = g.cluster:build_and_add_server({
+        alias = 'server_1',
+        box_cfg = g.box_cfg,
+    })
+    -- Servers 2 and 3 are interconnected without a proxy.
+    for i = 2, num_servers do
+        g.box_cfg.replication[i] = server.build_instance_uri('server_' .. i)
+    end
+    for i = 2, num_servers do
+        g.servers[i] = g.cluster:build_and_add_server({
+            alias = 'server_' .. i,
+            box_cfg = g.box_cfg,
+        })
+    end
+    g.proxies = {}
+    for i = 1, num_servers do
+        g.proxies[i] = proxy:new({
+            client_socket_path = server.build_instance_uri('server_' .. i .. '_proxy'),
+            server_socket_path = server.build_instance_uri('server_' .. i),
+        })
+        g.proxies[i]:start({force = true})
+    end
+    g.cluster:start()
+    g.cluster:wait_fullmesh()
+    g.servers[1]:exec(function()
+        box.schema.space.create('sync', {is_sync=true})
+        box.space.sync:create_index('pk')
+    end)
+    g.servers[2]:wait_vclock_of(g.servers[1])
+    g.servers[3]:wait_vclock_of(g.servers[1])
+end)
+
+g.after_all(function()
+    g.cluster:drop()
+end)
+
+g.test_wait_others = function(g)
```

Comment:
Nit: Masking `g` looks confusing. I think it's conventional to use `cg` for the local argument here.

Reply:
Ok.
<details><summary>diff</summary>

```diff
diff --git a/test/replication-luatest/linearizable_test.lua b/test/replication-luatest/linearizable_test.lua
index a342e1a4b..2e628e731 100644
--- a/test/replication-luatest/linearizable_test.lua
+++ b/test/replication-luatest/linearizable_test.lua
@@ -17,53 +17,53 @@ end
 
 local num_servers = 3
 
-g.before_all(function()
-    g.cluster = cluster:new({})
-    g.servers = {}
-    g.box_cfg = {
+g.before_all(function(cg)
+    cg.cluster = cluster:new({})
+    cg.servers = {}
+    cg.box_cfg = {
         replication = build_replication(num_servers),
         replication_timeout = 0.1,
         memtx_use_mvcc_engine = true,
     }
-    g.servers[1] = g.cluster:build_and_add_server({
+    cg.servers[1] = cg.cluster:build_and_add_server({
         alias = 'server_1',
-        box_cfg = g.box_cfg,
+        box_cfg = cg.box_cfg,
     })
     -- Servers 2 and 3 are interconnected without a proxy.
     for i = 2, num_servers do
-        g.box_cfg.replication[i] = server.build_instance_uri('server_' .. i)
+        cg.box_cfg.replication[i] = server.build_instance_uri('server_' .. i)
     end
     for i = 2, num_servers do
-        g.servers[i] = g.cluster:build_and_add_server({
+        cg.servers[i] = cg.cluster:build_and_add_server({
             alias = 'server_' .. i,
-            box_cfg = g.box_cfg,
+            box_cfg = cg.box_cfg,
         })
     end
-    g.proxies = {}
+    cg.proxies = {}
     for i = 1, num_servers do
-        g.proxies[i] = proxy:new({
+        cg.proxies[i] = proxy:new({
             client_socket_path = server.build_instance_uri('server_' .. i .. '_proxy'),
             server_socket_path = server.build_instance_uri('server_' .. i),
         })
-        g.proxies[i]:start({force = true})
+        cg.proxies[i]:start({force = true})
     end
-    g.cluster:start()
-    g.cluster:wait_fullmesh()
-    g.servers[1]:exec(function()
+    cg.cluster:start()
+    cg.cluster:wait_fullmesh()
+    cg.servers[1]:exec(function()
         box.schema.space.create('sync', {is_sync=true})
         box.space.sync:create_index('pk')
     end)
-    g.servers[2]:wait_vclock_of(g.servers[1])
-    g.servers[3]:wait_vclock_of(g.servers[1])
+    cg.servers[2]:wait_vclock_of(cg.servers[1])
+    cg.servers[3]:wait_vclock_of(cg.servers[1])
 end)
 
-g.after_all(function()
-    g.cluster:drop()
+g.after_all(function(cg)
+    cg.cluster:drop()
 end)
 
-g.test_wait_others = function(g)
-    g.proxies[1]:pause()
-    local fid = g.servers[1]:exec(function()
+g.test_wait_others = function(cg)
+    cg.proxies[1]:pause()
+    local fid = cg.servers[1]:exec(function()
         local t = require('luatest')
         local fiber = require('fiber')
         local f = fiber.new(function()
@@ -74,8 +74,8 @@ g.test_wait_others = function(g)
         t.assert_equals(f:status(), 'suspended', 'begin() waits for replicas')
         return f:id()
     end)
-    g.proxies[1]:resume()
-    local ok = g.servers[1]:exec(function(fid)
+    cg.proxies[1]:resume()
+    local ok = cg.servers[1]:exec(function(fid)
         local fiber = require('fiber')
         local _, ok, err =  fiber.join(fiber.find(fid))
         return ok, err
@@ -83,9 +83,9 @@ g.test_wait_others = function(g)
     t.assert(ok, 'begin() succeeds after replica connections are established')
 end
 
-g.test_timeout = function(g)
+g.test_timeout = function(cg)
     misc.skip_if_not_debug()
-    g.servers[1]:exec(function()
+    cg.servers[1]:exec(function()
         local t = require('luatest')
         box.error.injection.set('ERRINJ_RELAY_FROM_TX_DELAY', true)
         local ok, err = pcall(box.begin, {txn_isolation = 'linearizable',
@@ -97,37 +97,37 @@ g.test_timeout = function(g)
     end)
 end
 
-g.before_test('test_no_dirty_reads', function(g)
+g.before_test('test_no_dirty_reads', function(cg)
     -- Make cluster size 5. Having quorum 3 out of 5 will make the instance
     -- poll 5 - 3 + 1 == all 3 of the "real" nodes.
-    g.servers[1]:exec(function()
+    cg.servers[1]:exec(function()
         local uuid = require('uuid')
         box.space._cluster:insert{4, tostring(uuid.new())}
         box.space._cluster:insert{5, tostring(uuid.new())}
     end)
-    g.servers[2]:wait_vclock_of(g.servers[1])
-    g.servers[2]:exec(function()
+    cg.servers[2]:wait_vclock_of(cg.servers[1])
+    cg.servers[2]:exec(function()
         box.ctl.promote()
     end)
-    g.servers[1]:wait_vclock_of(g.servers[2])
-    g.servers[3]:wait_vclock_of(g.servers[2])
+    cg.servers[1]:wait_vclock_of(cg.servers[2])
+    cg.servers[3]:wait_vclock_of(cg.servers[2])
 end)
 
-g.after_test('test_no_dirty_reads', function(g)
-    g.servers[2]:exec(function()
+g.after_test('test_no_dirty_reads', function(cg)
+    cg.servers[2]:exec(function()
         box.ctl.demote()
     end)
-    g.servers[1]:wait_vclock_of(g.servers[2])
-    g.servers[1]:exec(function()
+    cg.servers[1]:wait_vclock_of(cg.servers[2])
+    cg.servers[1]:exec(function()
         box.space._cluster:delete{4}
         box.space._cluster:delete{5}
     end)
-    g.servers[2]:wait_vclock_of(g.servers[1])
-    g.servers[3]:wait_vclock_of(g.servers[1])
+    cg.servers[2]:wait_vclock_of(cg.servers[1])
+    cg.servers[3]:wait_vclock_of(cg.servers[1])
 end)
 
-g.test_no_dirty_reads = function(g)
-    g.replication_synchro_quorum = g.servers[2]:exec(function()
+g.test_no_dirty_reads = function(cg)
+    cg.replication_synchro_quorum = cg.servers[2]:exec(function()
         local t = require('luatest')
         local lsn = box.info.lsn
         local q = box.cfg.replication_synchro_quorum
@@ -136,7 +136,7 @@ g.test_no_dirty_reads = function(g)
         t.helpers.retrying({}, function() assert(box.info.lsn > lsn) end)
         return q
     end)
-    g.servers[1]:exec(function()
+    cg.servers[1]:exec(function()
         local t = require('luatest')
         local ok, err = pcall(box.begin, {txn_isolation = 'linearizable',
                                           timeout = 0.1})
@@ -146,10 +146,10 @@ g.test_no_dirty_reads = function(g)
                         'Pending transaction is seen')
         t.assert_equals(box.space.sync:get{1}, nil, 'No dirty reads')
     end)
-    g.servers[2]:exec(function(q)
+    cg.servers[2]:exec(function(q)
         box.cfg{replication_synchro_quorum = q}
-    end, {g.replication_synchro_quorum})
-    g.servers[1]:exec(function()
+    end, {cg.replication_synchro_quorum})
+    cg.servers[1]:exec(function()
         local t = require('luatest')
         box.begin{txn_isolation = 'linearizable'}
         local len = box.info.synchro.queue.len
@@ -163,15 +163,15 @@ g.test_no_dirty_reads = function(g)
     end)
 end
 
-g.test_reconnect_while_waiting = function(g)
+g.test_reconnect_while_waiting = function(cg)
     for i = 2, 3 do
-        g.servers[i]:exec(function()
+        cg.servers[i]:exec(function()
             local repl = table.copy(box.cfg.replication)
             table.remove(repl, 1)
             box.cfg{replication = repl}
         end)
     end
-    local f = fiber.new(g.servers[1].exec, g.servers[1], function()
+    local f = fiber.new(cg.servers[1].exec, cg.servers[1], function()
         local ok, err = pcall(box.begin, {txn_isolation = 'linearizable'})
         if ok then
             box.commit()
@@ -180,29 +180,29 @@ g.test_reconnect_while_waiting = function(g)
     end)
     f:set_joinable(true)
     for i = 2, 3 do
-        g.servers[i]:exec(function(repl)
+        cg.servers[i]:exec(function(repl)
             box.cfg{replication = repl}
-        end, {g.box_cfg.replication})
+        end, {cg.box_cfg.replication})
     end
     local success, ok, _ = f:join()
     t.assert(success, 'fiber joined successfully')
     t.assert(ok, 'reconnect while begin() waits for vclocks is tolerated')
 end
 
-g.test_leader_change = function(g)
-    g.servers[1]:exec(function()
+g.test_leader_change = function(cg)
+    cg.servers[1]:exec(function()
         box.ctl.promote()
     end)
-    g.servers[2]:wait_vclock_of(g.servers[1])
-    g.servers[3]:wait_vclock_of(g.servers[1])
+    cg.servers[2]:wait_vclock_of(cg.servers[1])
+    cg.servers[3]:wait_vclock_of(cg.servers[1])
     for i = 1, num_servers do
-        g.proxies[i]:pause()
+        cg.proxies[i]:pause()
     end
-    g.servers[2]:exec(function()
+    cg.servers[2]:exec(function()
         box.ctl.promote()
         box.space.sync:insert{2}
     end)
-    local fid = g.servers[1]:exec(function()
+    local fid = cg.servers[1]:exec(function()
         local f = require('fiber').new(function()
             box.begin{txn_isolation = 'linearizable'}
             local owner = box.info.synchro.queue.owner
@@ -216,31 +216,31 @@ g.test_leader_change = function(g)
         return f:id()
     end)
     for i = 1, num_servers do
-        g.proxies[i]:resume()
+        cg.proxies[i]:resume()
     end
-    local owner, get = g.servers[1]:exec(function(fid)
+    local owner, get = cg.servers[1]:exec(function(fid)
         local fiber = require('fiber')
         local _, owner, get = fiber.join(fiber.find(fid))
         return owner, get
     end, {fid})
-    t.assert_equals(owner, g.servers[2]:instance_id(),
+    t.assert_equals(owner, cg.servers[2]:instance_id(),
                     'leader change is noticed in box.begin()')
     t.assert_equals(get, {2}, 'Transaction committed by new leader is seen')
 end
 
 local g_basic = t.group('linearizable-read-basic')
 
-g_basic.before_all(function(g)
-    g.cluster = cluster:new({})
-    g.nomvcc = g.cluster:build_and_add_server({alias = 'nomvcc'})
-    g.mvcc = g.cluster:build_and_add_server({
+g_basic.before_all(function(cg)
+    cg.cluster = cluster:new({})
+    cg.nomvcc = cg.cluster:build_and_add_server({alias = 'nomvcc'})
+    cg.mvcc = cg.cluster:build_and_add_server({
         alias = 'mvcc',
         box_cfg = {
             memtx_use_mvcc_engine = true,
         },
     })
-    g.cluster:start()
-    g.mvcc:exec(function()
+    cg.cluster:start()
+    cg.mvcc:exec(function()
         box.schema.space.create('async'):create_index('pk')
         box.schema.space.create('sync', {is_sync = true}):create_index('pk')
         box.schema.space.create('local', {is_local = true}):create_index('pk')
@@ -253,8 +253,8 @@ g_basic.before_all(function(g)
     end)
 end)
 
-g_basic.after_all(function(g)
-    g.cluster:drop()
+g_basic.after_all(function(cg)
+    cg.cluster:drop()
 end)
 
 local function begin(opts)
@@ -291,24 +291,24 @@ local function test_linearizable_access(node, spacename, is_ok, msg)
     end
 end
 
-g_basic.test_basic = function(g)
+g_basic.test_basic = function(cg)
     local ok, err = pcall(box.cfg, {txn_isolation = 'linearizable'})
     t.assert(not ok, 'Cannot set default isolation level to linearizable')
     local msg = "Incorrect value for option 'txn_isolation': cannot set " ..
                 "default transaction isolation to 'linearizable'"
     t.assert_equals(err.message, msg,
                     'Correct error when trying to set default isolation')
-    ok, err = g.nomvcc:exec(begin, {{txn_isolation = 'linearizable'}})
+    ok, err = cg.nomvcc:exec(begin, {{txn_isolation = 'linearizable'}})
     t.assert(not ok, 'Error without mvcc enabled')
     t.assert_equals(err.message, 'Linearizable transaction does not support ' ..
                                  'disabled memtx mvcc engine',
                     'Correct error without mvcc enabled')
 
     local errmsg = 'space "async" does not support linearizable operations'
-    test_linearizable_access(g.mvcc, 'async', false, errmsg)
+    test_linearizable_access(cg.mvcc, 'async', false, errmsg)
     errmsg = 'space "vinyl" does not support linearizable operations'
-    test_linearizable_access(g.mvcc, 'vinyl', false, errmsg)
-    test_linearizable_access(g.mvcc, 'temporary', true)
-    test_linearizable_access(g.mvcc, 'local', true)
-    test_linearizable_access(g.mvcc, 'sync', true)
+    test_linearizable_access(cg.mvcc, 'vinyl', false, errmsg)
+    test_linearizable_access(cg.mvcc, 'temporary', true)
+    test_linearizable_access(cg.mvcc, 'local', true)
+    test_linearizable_access(cg.mvcc, 'sync', true)
 end
```
</details>